python/astro_metadata_translator/translator.py

```diff
@@ -943,6 +943,39 @@ def to_observation_counter(self):
         """
         return 0
 
+    @classmethod
+    def read_all_headers(cls, filename, primary=None):
+        """Read all relevant headers from the given file.
+
+        This should simply return all headers, but for multi-extension FITS
+        files where each file includes data from multiple detectors,
+        all the headers for each detector should be returned.
+
+        Parameters
+        ----------
+        filename : `str`
+            Path to a file in a format understood by this translator.
+        primary : `dict`-like, optional
+            The primary header obtained by the caller. This is sometimes
+            already known, for example if a system is trying to bootstrap
+            without already knowing what data is in the file. For many
+            instruments where the primary header is the only relevant
+            header, the primary header will be returned with no further
+            action.
+
+        Yields
+        ------
+        headers : iterator of `dict`-like
+            Each relevant header in turn.
+        """
+        if primary is not None:
+            yield primary
+        else:
```

Comment:
That is intended. In the general case if the routine is given the primary header it's unlikely that this routine can guess well enough to improve on the situation. I was trying to make it simple. Would you prefer that I look for an INHERIT=T header and EXTEND=T header and by default try to read the second header and merge it? I don't think it can do anything in the default case other than return one header (merged or not) since for an arbitrary FITS file of unknown origin I can't tell if subsequent HDUs are (say) amplifiers or detector images.

Reply:
The other point here is that I don't want to be forced into reading every header twice. I have to read it once in order to work out which metadata translator class to call. In many cases that header that I read is sufficient to do metadata translation so it's much better to return it immediately rather than require that we then read the same header a second time (or possibly third time if we merged in the caller because it was a compressed image like those in testdata_ci_hsc).

Reply:
Actually, what I had in mind was not having a default implementation at all, but then you'd need a `SingleHeaderTranslator` subclass that everything other than LSST/DECam/CHFT use.

If you keep the code organized as-is, I recommend at least including a note in the docstring to the effect of "This implementation returns `primary`, if provided, otherwise it reads the first header from `filename`. Translator subclasses for instruments with additional headers must override this method." Otherwise, subclass authors have no way of knowing that supporting multiple headers is not the base class's responsibility.

Reply:
Okay. I really don't want to have to insert a new class into the hierarchy that will be the default so I will make it clearer in the documentation what the default implementation will do.

Reply:
`read_translatable_heaers` ?

Reply:
I have completely revamped the documentation and changed the method name to `determine_translatable_headers`